packages/functional/src/curry.ts

```diff
@@ -0,0 +1,382 @@
+    a: A,
+    b: B,
+    c: C,
+    d: D,
+    e: E,
+    f: F,
+    g: G,
+    h: H
+): (i: I) => R;
+export function curry<A, B, C, D, E, F, G, H, I, R>(
+    fn: (a: A, b: B, c: C, d: D, e: E, f: F, g: G, h: H, i: I) => R,
+    a: A,
+    b: B,
+    c: C,
+    d: D,
+    e: E,
+    f: F,
+    g: G,
+    h: H,
+    i: I
+): () => R;
+// 10 args
+export function curry<A, B, C, D, E, F, G, H, I, J, R>(
+    fn: (a: A, b: B, c: C, d: D, e: E, f: F, g: G, h: H, i: I, j: J) => R
+): (a: A, b: B, c: C, d: D, e: E, f: F, g: G, h: H, i: I, j: J) => R;
+export function curry<A, B, C, D, E, F, G, H, I, J, R>(
+    fn: (a: A, b: B, c: C, d: D, e: E, f: F, g: G, h: H, i: I, j: J) => R,
+    a: A
+): (b: B, c: C, d: D, e: E, f: F, g: G, h: H, i: I, j: J) => R;
+export function curry<A, B, C, D, E, F, G, H, I, J, R>(
+    fn: (a: A, b: B, c: C, d: D, e: E, f: F, g: G, h: H, i: I, j: J) => R,
+    a: A,
+    b: B
+): (c: C, d: D, e: E, f: F, g: G, h: H, i: I, j: J) => R;
+export function curry<A, B, C, D, E, F, G, H, I, J, R>(
+    fn: (a: A, b: B, c: C, d: D, e: E, f: F, g: G, h: H, i: I, j: J) => R,
+    a: A,
+    b: B,
+    c: C
+): (d: D, e: E, f: F, g: G, h: H, i: I, j: J) => R;
+export function curry<A, B, C, D, E, F, G, H, I, J, R>(
+    fn: (a: A, b: B, c: C, d: D, e: E, f: F, g: G, h: H, i: I, j: J) => R,
+    a: A,
+    b: B,
+    c: C,
+    d: D
+): (e: E, f: F, g: G, h: H, i: I, j: J) => R;
+export function curry<A, B, C, D, E, F, G, H, I, J, R>(
+    fn: (a: A, b: B, c: C, d: D, e: E, f: F, g: G, h: H, i: I, j: J) => R,
+    a: A,
+    b: B,
+    c: C,
+    d: D,
+    e: E
+): (f: F, g: G, h: H, i: I, j: J) => R;
+export function curry<A, B, C, D, E, F, G, H, I, J, R>(
+    fn: (a: A, b: B, c: C, d: D, e: E, f: F, g: G, h: H, i: I, j: J) => R,
+    a: A,
+    b: B,
+    c: C,
+    d: D,
+    e: E,
+    f: F
+): (g: G, h: H, i: I, j: J) => R;
+export function curry<A, B, C, D, E, F, G, H, I, J, R>(
+    fn: (a: A, b: B, c: C, d: D, e: E, f: F, g: G, h: H, i: I, j: J) => R,
+    a: A,
+    b: B,
+    c: C,
+    d: D,
+    e: E,
+    f: F,
+    g: G
+): (h: H, i: I, j: J) => R;
+export function curry<A, B, C, D, E, F, G, H, I, J, R>(
+    fn: (a: A, b: B, c: C, d: D, e: E, f: F, g: G, h: H, i: I, j: J) => R,
+    a: A,
+    b: B,
+    c: C,
+    d: D,
+    e: E,
+    f: F,
+    g: G,
+    h: H
+): (i: I, j: J) => R;
+export function curry<A, B, C, D, E, F, G, H, I, J, R>(
+    fn: (a: A, b: B, c: C, d: D, e: E, f: F, g: G, h: H, i: I, j: J) => R,
+    a: A,
+    b: B,
+    c: C,
+    d: D,
+    e: E,
+    f: F,
+    g: G,
+    h: H,
+    i: I
+): (j: J) => R;
+export function curry<A, B, C, D, E, F, G, H, I, J, R>(
+    fn: (a: A, b: B, c: C, d: D, e: E, f: F, g: G, h: H, i: I, j: J) => R,
+    a: A,
+    b: B,
+    c: C,
+    d: D,
+    e: E,
+    f: F,
+    g: G,
+    h: H,
+    i: I,
+    j: J
+): () => R;
+// Implementation
+export function curry(fn: (...args: unknown[]) => unknown, ...args: unknown[]) {
+    return fn.bind(null, ...args);
```

Comment:
Couple notes about this.

* Currying hipsters are used to being able to chain calls like this: `sum3(1)(2)(3)`
* Currying hipsters are also used to the functions auto-invoking once you've met or exceeded the arity of the original function: `sum3(1)(2, 3); // 6`
* It would be nice in any case to have the property that `curry()` either invokes immediately, or returns the original function when the thing has no effect, so that we're not needlessly creating copies
    ```ts
    const f = () => {};
    const fCurried = curry(f);
    f === fCurried; // true
    ````

Reply:
OK, I can add this behavior, as long as we're in favor of adding it to the lib. Sounds like @mcintyre94 was on the fence, and I am in favor. Steve is... ?

Reply:
I think I'm also sort of on the fence. I haven't needed to do this yet, at least not in quite the way that Real Currying™ allows.